shpc/main/modules/module.py

```diff
@@ -97,6 +86,49 @@ def container_path(self):
             )
         return self._container_path
 
+    def add_local_container(self, container_image):
+        """
+        Set the module container to be a local container (copied over)
+        """
+        basename = os.path.basename(container_image)
+        container_dest = os.path.join(self.container_dir, basename)
+        if not os.path.exists(self.container_dir):
+            utils.mkdir_p(self.container_dir)
+        if not os.path.exists(container_dest):
+            shutil.copyfile(container_image, container_dest)
```

Comment:
I can see pros and cons for both doing a copy (shpc remains in control of its containers, but it costs extra disk space) vs a symlink (no duplication on disk, but risk that the source file will go away.

I think I was expecting a symlink here ... Could this be an option ?
@audreystott : what behaviour would you expect ?

Reply:
ping @audreystott - it sounds like you'd want to simply use the original path (without copying) so I'll figure out how to add a setting for that.

Reply:
Yep that’s right! All of the biocontainers are already pulled into that server and named, so SHPC won’t have to pull them.

Reply:
okay - just added support for `--keep-path` which should install the container as is. Next we need a matching of the registry containers with local files - you can likely pick one out to test locally with this branch! I'm off to bed, back tomorrow.

Reply:
Hmm. It still creates the new sif file. (I am on the correct branch and on this PR).

I did this:
```
$ shpc add /cvmfs/singularity.galaxyproject.org/all/fastqc:0.11.9--0 fastqc
```
And the result is the same container.yaml file:
```
# change this to a URL to describe your container, or to get help
url: https://singularity-hpc.readthedocs.io

# change this to your GitHub alias (or another contact or name)
maintainer: Dinosaur

# Any custom features?
# features:
#  gpu: true

# Put custom aliases here
# aliases:
#  echo: /usr/bin/echo

# custom environment variables
# env:
#  breakfast: pancakes

# A description to describe your container
description: A custom container to do X.
latest:
  latest: sha256:56303eba371672d61c9a497702c207e4e47a9c8766d1edc3d9d75351f00332f5
tags:
  0.11.9--0: sha256:56303eba371672d61c9a497702c207e4e47a9c8766d1edc3d9d75351f00332f5
  latest: sha256:56303eba371672d61c9a497702c207e4e47a9c8766d1edc3d9d75351f00332f5
path: sha256:56303eba371672d61c9a497702c207e4e47a9c8766d1edc3d9d75351f00332f5.sif
```
And hence same result with installing:
```
$ shpc install fastqc:0.11.9--0
Error creating path /cvmfs/singularity.galaxyproject.org/all/fastqc/0.11.9--0, exiting.
```

Reply:
@audreystott I think you are in a different (completely opposite!) timezone, so let me know when you give the above a try! This should only be generating files in your module directory, which can be separate from where the containers are. If this works without a hitch we should be good to merge here, and tonight I'll try to finish up work to update the registry for the 8k containers, and then likely tomorrow / later this week come up with a script (or strategy) to install a latest based on the containers you find (and what is available in the registry!)

Apologies there are a lot of pieces but I haven't forgotten your use case!

Reply:
@vsoch sorry, typical user not reading the doco properly. I must be doing something wrong as it's giving the same error. 

Just to check, is it correct that I use my own registry and I do `shpc add /cvmfs/singularity.galaxyproject.org/all/fastqc:0.11.9--0 fastqc:0.11.9--0` and then `shpc install fastqc:0.11.9--0 /cvmfs/singularity.galaxyproject.org/fastqc:0.11.9--0 --keep-path`?

```
$ shpc add /cvmfs/singularity.galaxyproject.org/all/fastqc:0.11.9--0 fastqc:0.11.9--0
fastqc:0.11.9--0 already exists and will be updated!
Registry entry fastqc was added! Before shpc install, edit:
/home/ubuntu/shpc/registry/fastqc/container.yaml

$ shpc install fastqc:0.11.9--0 /cvmfs/singularity.galaxyproject.org/fastqc:0.11.9--0 --keep-path
Error creating path /cvmfs/singularity.galaxyproject.org/all/fastqc/0.11.9--0, exiting.
```
My timezone is UTC+8.

Reply:
No it’s not - look at the new docs here, you just want the command that I showed you, an install with a container identifier, then the SIF, then the flag. You shouldn’t be running ads at all!

Reply:
Also make sure you delete the module first I’d you accidentally installed something.

Reply:
Which registry is it supposed to look at? The quay.io/biocontainers one?

Reply:
If using the shpc-registry one, `registry: [/home/ubuntu/shpc-registry/quay.io/biocontainers]`.

Then:
```
$ shpc install fastqc:0.11.9--0 /cvmfs/singularity.galaxyproject.org/all/fastqc:0.11.9--0 --keep-path
Error creating path /cvmfs/singularity.galaxyproject.org/all/quay.io/biocontainers/fastqc/0.11.9--0, exiting.
```

Reply:
Ah okay! So I think there must be a spot we missed where it tries to create the container directory (and we don't want that). I'm having dinner (and not wearing glasses!) but I'll take a look right after!

Reply:
:D All good. No rush!

Reply:
@audreystott could you run these commands to get the paths for your modules and containers to be installed and show them to me?

```
$ shpc config get module_base
module_base                    /home/vanessa/Desktop/Code/shpc/modules
$ shpc config get container_base
container_base                 /home/vanessa/Desktop/Code/shpc/containers
```
I'm suspicious it's trying to write to this random location with "all" - it shouldn't do that unless you've specific to write there. We will need a directory to write the modules, however, and we will need write there (and this can be separate from the containers).

Reply:
@vsoch 
```
$ shpc config get module_base
module_base                    /home/ubuntu/shpc/modules
$ shpc config get container_base
container_base                 /cvmfs/singularity.galaxyproject.org/all
```
I had it set to install to that container_base location, so that it would see that the container is already there.

Reply:
Ah, so how about unset that for now - it will default to the same as the module base (which I think is what we want).  Then again do:

```
$ shpc install fastqc:0.11.9--0 /cvmfs/singularity.galaxyproject.org/fastqc:0.11.9--0 --keep-path
```
and then see if the error still happens!

Reply:
You solved it!! The module is created and in there sits the correct path to the container!
```
Container (available through variable SINGULARITY_CONTAINER):

 - /cvmfs/singularity.galaxyproject.org/all/fastqc:0.11.9--0
 ```

```
$ shpc list
  quay.io/biocontainers/fastqc:0.11.9--0
```

Reply:
Hooray!! So give that a good testing, try to break it, and when you are happy I’ll merge and release here. I’m almost done with an action to incrementally update our (TBA) much larger container registry, so when that is merged I’ll follow up with the container.yaml files and then get started on some means to mass install!

Reply:
Amazing! I will go and test a whole bunch of it today. Please go and rest now!!